Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py

```diff
@@ -0,0 +1,578 @@
+        self.base_url = f'{url}/subscriptions/{subscription_id}/resourceGroups/{resource_group_name}/providers/' \
+            f'Microsoft.OperationalInsights/workspaces/{workspace_name}/providers/Microsoft.SecurityInsights'
+        self.tenant_id = tenant_id
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.auth_code = auth_code
+        self.access_token = self.get_access_token()
+        super(Client, self).__init__(self.base_url, **kwargs)
+
+    def get_access_token(self):
+        integration_context = demisto.getIntegrationContext()
+        access_token = integration_context.get('access_token')
+        refresh_token = integration_context.get('refresh_token')
+        access_token_expiration_time = dateparser.parse(integration_context.get('access_token_expiration_time', '0'))
+
+        if not access_token or datetime.now() - access_token_expiration_time > timedelta(seconds=-5):
+            access_token = self.make_access_token_request(refresh_token)
+
+        return access_token
+
+    def make_access_token_request(self, refresh_token=None, retry=False):
+        data = self.get_access_token_request_data(refresh_token)
+        res = requests.post(f'https://login.microsoftonline.com/{self.tenant_id}/oauth2/token',
+                            headers={'Content-Type': 'application/x-www-form-urlencoded'},
+                            data=data)
+
+        if res.status_code != 200:  # Refresh token has expired
+            if not retry:  # Try again using auth_code
+                return self.make_access_token_request(refresh_token=None, retry=True)
+            else:
+                raise Exception(AUTHORIZATION_ERROR_MSG)
+
+        try:
+            res_json = res.json()
+            self.update_tokens_in_context(res_json)
+            return res_json.get('access_token')
+        except ValueError:
+            raise Exception(AUTHORIZATION_ERROR_MSG)
+
+    def get_access_token_request_data(self, refresh_token=None):
+        data = {
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'redirect_uri': 'https://localhost/myapp',
+            'resource': 'https://management.core.windows.net'
+        }
+
+        if refresh_token:
+            data['grant_type'] = 'refresh_token'
+            data['refresh_token'] = refresh_token
+        else:
+            data['grant_type'] = 'authorization_code'
+            data['code'] = self.auth_code
+
+        return data
+
+    def update_tokens_in_context(self, res):
+        integration_context = {
+            'access_token': res.get('access_token'),
+            'refresh_token': res.get('refresh_token'),
+            'access_token_expiration_time': res.get('expires_on', '0')
+        }
+        demisto.setIntegrationContext(integration_context)
+
+    def http_request(self, method, url_suffix=None, full_url=None, params=None, data=None):
```

Comment:
why are we not using `BaseClient._http_request` ?

Reply:
http_request is a wrapper function for BaseClient._http_request(), I check here things specifically for this integration